docs/parameters.md
Waiter's behavior is customized through parameters specified by the HTTP client.

The basic parameters that Waiter supports are:

|Parameter|Required?|Default Value|Description|
|---------|---------|-------------|-----------|
|`X-Waiter-Cmd`|**Yes**|n/a|The shell command to run. Your application must listen on a port chosen by Waiter. The port your service should listen in is set via the `$PORT0` environment variable.|
|`X-Waiter-Cpus`|**Yes**|n/a|The number of cores your service needs.|
|`X-Waiter-Health-Check-Url`|No|"/status"|Specifies an HTTP endpoint that Waiter will hit to determine if your app is healthy. This should return `200 OK`.|
|`X-Waiter-Mem`|**Yes**|n/a|The amount of memory your service needs (MB). You will still need to explicitly include any memory flags (e.g. `-Xmx`) in your command.|
|`X-Waiter-Name`|No|empty string|A name for your service. Use only letters, numbers, dashes, underscores, and dots.|
|`X-Waiter-Run-As-User`|No|user who is making the request|The user that the service should be run as.|
|`X-Waiter-Version`|**Yes**|n/a|Specify a version to associate with your service (e.g. "a0b1c2d3e4f5" or "my-version-name").|

Additional (optional) parameters that can be set:

|Parameter|Default Value|Valid Values|Description|Guidance|
|---------|-------------|------------|-----------|--------|
|`X-Waiter-Blacklist-On-503`|true|true or false|If an instance returns 503, whether or not the instance will be blacklisted.|By default, Waiter avoids instances that are returning a 503 error code in responses, which typically indicates the server is too busy. If you would prefer that Waiter not do this, feel free to disable this feature.|
|`X-Waiter-Cmd-Type`|"shell"|"shell"|Provides an extension point for supporting different types of commands in the future.|Feel free to add new command types to suit your needs.|
|`X-Waiter-Concurrency-Level`|1|1-10000|The number of simultaneous requests to an individual backend instance.|Increasing this value will likely lead to better performance and better resource utilization. Avoid queuing your requests on your backend instances by not increasing concurrency level above what an individual instance can handle. The shorter your requests, the more benefit you will get out of a higher concurrency level.|
|`X-Waiter-Distribution-Scheme`|"balanced"|"balanced" or "simple"|The strategy used by Waiter routers to distribute instances amongst themselves. A "simple" distribution, achieved using consistent hashing, will minimize instances being assigned to multiple routers due to scaling. A "balanced" distribution ensures even splitting of instances across routers after the initial consistent hash distribution.|If you see too many blacklisted instances due to 503 responses, it may be a good idea to try out "simple" distribution.|
|`X-Waiter-Env`|none|A JSON map from string keys to string values|The keys and values correspond to environment variable names and values which will get passed to the command.|Use as needed.|
|`X-Waiter-Expired-Instance-Restart-Rate`|0.1|(0-1]|The max percentage of total expired instances to recycle at the same time (e.g. if the service has 10 expired instances and a `X-Waiter-Expired-Instance-Restart-Rate` of 0.1, Waiter will only restart 1 at a time).|The default is reasonable for most cases.|
|`X-Waiter-Grace-Period-Secs`|30|1-3600 (1 hour)|The amount of time to wait after a service starts before considering a health check failure to be an error.|Set this value to startup time of your service. If your service starts up within 30 seconds, don't bother with changing this value.|
|`X-Waiter-Idle-Timeout-Mins`|4320 (3 days)|1-10080 (1 week)|The number of minutes of no activity to wait before shutting down a service completely.|Setting a lower idle timeout is strongly encouraged and frees cluster resources for other services to use.|
|`X-Waiter-Instance-Expiry-Mins`|7200 (5 days)|>0|The amount of time to allow a particular instance of your app to run before restarting it. Waiter ensures another instance is available before killing the expiring instance such that service is unaffected.|You can set this value to be as large as you like.|
|`X-Waiter-Jitter-Threshold`|0.5|[0,1)|How big the difference between ideal and current number of instances must be for Waiter to scale your service.|The default is reasonable for most cases.|
|`X-Waiter-Max-Instances`|500|1-500|The maximum number of instances that Waiter should not scale above.|Setting a max can be helpful for preventing unexpected scaling.|
|`X-Waiter-Max-Queue-Length`|1000|>0|The number of requests to queue before rejecting requests with a 503 error code.|Lowering this value to something reasonable for your architecture is encouraged as it allows you to fail faster. Don't increase the value without a good reason.|
|`X-Waiter-Metadata-*`|n/a|any string|A metadata field. Metadata fields allow you to store arbitrary strings along with your service description.|Allows additional information to be stored with a service to help identify the service.|
|`X-Waiter-Metric-Group`|"other"|[A-Za-z0-9-_]+|Metric groups allow services to be grouped together for the purpose of metric collection.|If you have `:statsd` enabled in your [config file](../resources/default-config.clj), then specify a metric group. If you have multiple services, choose the same metric group if you'd like combined metrics, otherwise choose independent metric groups.|
|`X-Waiter-Min-Instances`|1|1 or 2|The minimum number of instances that Waiter should keep running at all times.|Go with the default.|
|`X-Waiter-Permitted-User`|user who is making the request|any user|The user that is authorized to make requests to your service|Use as needed.|
|`X-Waiter-Queue-Timeout`|300000 (5 mins)|>0|The maximum time, in milliesconds, allowed spent by a request waiting for an available service backend.|Don't change it unless you have a good reason to do so.|
|`X-Waiter-Restart-Backoff-Factor`|2|>=1|The factor by which to multiple the amount of time between consecutive start-up failures. The base is 1 second.|The default is reasonable for most cases.|
|`X-Waiter-Scale-Down-Factor`|0.1|(0-1]|The percentage amount, per second, to decrease the current number of instances toward the target number of instances. See [How Waiter Autoscaling Works](autoscaling.md).|Don't change it unless you have a good reason to do so.|
|`X-Waiter-Scale-Factor`|1|(0-1]|The ratio of desired number of backends to number of outstanding requests. See [How Waiter Autoscaling Works](autoscaling.md).|Don't change it unless you have a good reason to do so, although values lower than 1 promote more efficient usage of resources if you don't mind waiting.|
|`X-Waiter-Scale-Up-Factor`|0.01|(0-1]|The percentage amount, per second, to increase the current number of instances toward the target number of instances. See [How Waiter Autoscaling Works](autoscaling.md).|Don't change it unless you have a good reason to do so.|
|`X-Waiter-Timeout`|900000 (15 mins)|>0|The socket timeout, in milliesconds, of the connection between Waiter and a service backend.|Don't change it unless you have a good reason to do so.|


waiter/src/waiter/settings.clj
;;
;;       Copyright (c) 2017 Two Sigma Investments, LLC.
;;       All Rights Reserved
;;
;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
;;       Two Sigma Investments, LLC.
;;
;;       The copyright notice above does not evidence any
;;       actual or intended publication of such source code.
;;
(ns waiter.settings
  (:require [clojure.edn :as edn]
            [clojure.java.io :as io]
            [clojure.tools.logging :as log]
            [schema.core :as s]
            [waiter.schema :as schema]
            [waiter.utils :as utils]))

(def settings-schema
  {(s/required-key :authenticator-config) (s/constrained
                                            {:kind s/Keyword
                                             s/Keyword schema/require-symbol-factory-fn}
                                            schema/contains-kind-sub-map?)
   (s/required-key :blacklist-config) {(s/required-key :blacklist-backoff-base-time-ms) schema/positive-int
                                       (s/required-key :max-blacklist-time-ms) schema/positive-int}
   (s/required-key :cors-config) (s/constrained
                                   {:kind s/Keyword
                                    (s/optional-key :ttl) schema/positive-int
                                    (s/required-key :max-age) schema/positive-int
                                    s/Keyword schema/require-symbol-factory-fn}
                                   schema/contains-kind-sub-map?)
   (s/required-key :cluster-config) {(s/required-key :min-routers) schema/positive-int
                                     (s/required-key :name) schema/non-empty-string}
   (s/required-key :consent-expiry-days) schema/positive-int
   (s/required-key :entitlement-config) (s/constrained
                                          {:kind s/Keyword
                                           (s/optional-key :cache) {(s/required-key :threshold) schema/positive-int
                                                                    (s/required-key :ttl) schema/positive-int}
                                           s/Keyword schema/require-symbol-factory-fn}
                                          schema/contains-kind-sub-map?)
   (s/optional-key :git-version) s/Any
   (s/required-key :health-check-timeout-ms) schema/positive-int
   (s/required-key :host) schema/non-empty-string
   (s/required-key :hostname) schema/non-empty-string
   (s/required-key :instance-request-properties) {(s/required-key :async-check-interval-ms) schema/positive-int
                                                  (s/required-key :async-request-timeout-ms) schema/positive-int
                                                  (s/required-key :connection-timeout-ms) schema/positive-int
                                                  (s/required-key :initial-socket-timeout-ms) schema/positive-int
                                                  (s/required-key :streaming-timeout-ms) schema/positive-int
                                                  (s/required-key :queue-timeout-ms) schema/positive-int}
   (s/required-key :kv-config) (s/constrained
                                 {:kind s/Keyword
                                  (s/optional-key :encrypt) s/Bool
                                  (s/optional-key :cache) {(s/required-key :threshold) schema/positive-int
                                                           (s/required-key :ttl) schema/positive-int}
                                  s/Keyword schema/require-symbol-factory-fn}
                                 schema/contains-kind-sub-map?)
   (s/optional-key :messages) {s/Keyword s/Str}
   (s/required-key :metric-group-mappings) schema/valid-metric-group-mappings
   (s/required-key :metrics-config) {(s/required-key :inter-router-metrics-idle-timeout-ms) schema/positive-int
                                     (s/required-key :metrics-gc-interval-ms) schema/positive-int
                                     (s/required-key :metrics-sync-interval-ms) schema/positive-int
                                     (s/required-key :router-update-interval-ms) schema/positive-int
                                     (s/required-key :transient-metrics-timeout-ms) schema/positive-int}
   (s/required-key :password-store-config) (s/constrained
                                             {:kind s/Keyword
                                              s/Keyword schema/require-symbol-factory-fn}
                                             schema/contains-kind-sub-map?)
   (s/required-key :port) schema/positive-int
   (s/required-key :router-id-prefix) s/Str
   (s/required-key :router-syncer) {(s/required-key :delay-ms) schema/positive-int
                                    (s/required-key :interval-ms) schema/positive-int}
   (s/required-key :scaling) {(s/required-key :autoscaler-interval-ms) schema/positive-int
                              (s/required-key :inter-kill-request-wait-time-ms) schema/positive-int}
   (s/required-key :scheduler-config) (s/constrained
                                        {:kind s/Keyword
                                         s/Keyword schema/require-symbol-factory-fn}
                                        schema/contains-kind-sub-map?)
   (s/required-key :scheduler-gc-config) {(s/required-key :broken-service-min-hosts) schema/positive-int
                                          (s/required-key :broken-service-timeout-mins) schema/positive-int
                                          (s/required-key :scheduler-gc-broken-service-interval-ms) schema/positive-int
                                          (s/required-key :scheduler-gc-interval-ms) schema/positive-int}
   (s/required-key :scheduler-syncer-interval-secs) schema/positive-int
   (s/required-key :service-description-builder-config) (s/constrained
                                                          {:kind s/Keyword
                                                           s/Keyword schema/require-symbol-factory-fn}
                                                          schema/contains-kind-sub-map?)
   ; service-description-defaults should never contain default values for required fields, e.g. version, cmd, run-as-user, etc.
   (s/required-key :service-description-defaults) {(s/required-key "blacklist-on-503") s/Bool
                                                   (s/required-key "concurrency-level") schema/positive-int
                                                   (s/required-key "distribution-scheme") (s/enum "balanced" "simple")
                                                   (s/required-key "env") {s/Str s/Str}
                                                   (s/required-key "expired-instance-restart-rate") schema/positive-fraction-less-than-or-equal-to-1
                                                   (s/required-key "grace-period-secs") schema/positive-int
                                                   (s/required-key "health-check-url") schema/non-empty-string
                                                   (s/required-key "idle-timeout-mins") schema/positive-int
                                                   (s/required-key "instance-expiry-mins") schema/positive-int
                                                   (s/required-key "jitter-threshold") schema/greater-than-or-equal-to-0-less-than-1
                                                   (s/required-key "max-instances") schema/positive-int
                                                   (s/required-key "max-queue-length") schema/positive-int
                                                   (s/required-key "metadata") {s/Str s/Str}
                                                   (s/required-key "min-instances") schema/positive-int
                                                   (s/required-key "permitted-user") schema/non-empty-string
                                                   (s/required-key "restart-backoff-factor") schema/positive-number-greater-than-or-equal-to-1
                                                   (s/required-key "scale-factor") schema/positive-fraction-less-than-or-equal-to-1
                                                   (s/required-key "scale-up-factor") schema/positive-fraction-less-than-1
                                                   (s/required-key "scale-down-factor") schema/positive-fraction-less-than-1}
   (s/required-key :statsd) (s/either (s/eq :disabled)
                                      {(s/required-key :cluster) schema/non-empty-string
                                       (s/required-key :environment) schema/non-empty-string
                                       (s/optional-key :histogram-max-size) schema/positive-int
                                       (s/required-key :host) schema/non-empty-string
                                       (s/required-key :port) schema/positive-int
                                       (s/required-key :publish-interval-ms) schema/positive-int
                                       (s/required-key :server) schema/non-empty-string})
   (s/required-key :thread-stack-state-refresh-interval-ms) schema/positive-int
   (s/required-key :work-stealing) {(s/required-key :offer-help-interval-ms) schema/positive-int
                                    (s/required-key :reserve-timeout-ms) schema/positive-int}
   (s/required-key :zookeeper) {(s/required-key :base-path) schema/non-empty-string
                                (s/required-key :connect-string) schema/valid-zookeeper-connect-config
                                (s/required-key :curator-retry-policy) {(s/required-key :base-sleep-time-ms) schema/positive-int
                                                                        (s/required-key :max-sleep-time-ms) schema/positive-int
                                                                        (s/required-key :max-retries) schema/positive-int}
                                (s/required-key :discovery-relative-path) schema/non-empty-string
                                (s/required-key :gc-relative-path) schema/non-empty-string
                                (s/required-key :leader-latch-relative-path) schema/non-empty-string
                                (s/required-key :mutex-timeout-ms) schema/positive-int}})

(defn load-settings-file
  "Loads the edn config in the specified file, it relies on having the filename being a path to the file."
  [filename]
  (let [config-file (-> filename str io/file)
        config-file-path (.getAbsolutePath config-file)]
    (if (.exists config-file)
      (do
        (log/info "reading settings from file:" config-file-path)
        (let [edn-readers {:readers {'regex (fn [expr] (re-pattern expr))}}
              settings (edn/read-string edn-readers (slurp config-file-path))]
          (log/info "configured settings:\n" (with-out-str (clojure.pprint/pprint settings)))
          settings))
      (do
        (log/info "unable to find configuration file:" config-file-path)
        (utils/exit 1 (str "Unable to find configuration file: " config-file-path))))))

(defn sanitize-settings
  "Sanitizes settings for eventual conversion to JSON"
  [settings]
  (utils/dissoc-in settings [:zookeeper :connect-string]))

(defn display-settings
  "Endpoint to display the current settings in use."
  [settings]
  (utils/map->json-response (into (sorted-map) (sanitize-settings settings))))

(def settings-defaults
  {:authenticator-config {:kind :one-user
                          :one-user {:factory-fn 'waiter.auth.authentication/one-user-authenticator}}
   :cors-config {:kind :patterns
                 :patterns {:factory-fn 'waiter.cors/pattern-based-validator
                            :allowed-origins []}
                 :max-age 3600}
   :blacklist-config {:blacklist-backoff-base-time-ms 10000
                      :max-blacklist-time-ms 300000}
   ;; To be considered part of the same cluster, routers need to
   ;; 1. have the same leader-latch-path to participate in leadership election
   ;; 2. have the same discovery path with the same cluster name to allow computing router endpoints
   :cluster-config {:min-routers 1
                    :name "waiter"}
   :consent-expiry-days 90
   :entitlement-config {:kind :simple
                        :simple {:factory-fn 'waiter.security/->SimpleEntitlementManager}}
   :health-check-timeout-ms 200
   :host "0.0.0.0"
   :hostname "localhost"
   :instance-request-properties {:async-check-interval-ms 3000
                                 :async-request-timeout-ms 60000
                                 :connection-timeout-ms 5000 ; 5 seconds
                                 :initial-socket-timeout-ms 900000 ; 15 minutes
                                 :queue-timeout-ms 300000
                                 :streaming-timeout-ms 20000}
   :kv-config {:kind :zk
               :zk {:factory-fn 'waiter.kv/new-zk-kv-store
                    :sync-timeout-ms 2000}
               :cache {:threshold 1000
                       :ttl 60}
               :encrypt true
               :relative-path "tokens"}
   :messages {:cannot-identify-service "Unable to identify service using waiter headers/token"
              :invalid-service-description "Service description using waiter headers/token improperly configured"}
   :metric-group-mappings []
   :metrics-config {:inter-router-metrics-idle-timeout-ms 2000
                    :metrics-gc-interval-ms 60000
                    :metrics-sync-interval-ms 50
                    :router-update-interval-ms 5000
                    :transient-metrics-timeout-ms 300000}
   :password-store-config {:kind :configured
                           :configured {:factory-fn 'waiter.password-store/configured-provider
                                        :passwords ["open-sesame"]}}
   :port 9091
   :router-id-prefix ""
   :router-syncer {:delay-ms 750
                   :interval-ms 1500}
   :scaling {:autoscaler-interval-ms 1000
             ; throttles the rate at which kill requests are sent to the scheduler
             :inter-kill-request-wait-time-ms 1000}
   :scheduler-config {:kind :marathon
                      :marathon {:factory-fn 'waiter.marathon/marathon-scheduler
                                 :home-path-prefix "/home/"
                                 :http-options {:conn-timeout 10000
                                                :socket-timeout 10000
                                                :spnego-auth true}
                                 :force-kill-after-ms 60000
                                 :framework-id-ttl 900000}
                      :shell {:factory-fn 'waiter.shell-scheduler/shell-scheduler
                              :health-check-interval-ms 10000
                              :health-check-timeout-ms 200
                              :port-grace-period-ms 120000
                              :port-range [10000 10999]
                              :work-directory "scheduler"}}
   :scheduler-gc-config {:broken-service-min-hosts 2
                         :broken-service-timeout-mins 30
                         :scheduler-gc-broken-service-interval-ms 60000
                         :scheduler-gc-interval-ms 60000}
   :scheduler-syncer-interval-secs 5
   :service-description-builder-config {:kind :default
                                        :default {:factory-fn
                                                  'waiter.service-description/->DefaultServiceDescriptionBuilder}}
   :service-description-defaults {"blacklist-on-503" true
                                  "concurrency-level" 1
                                  "distribution-scheme" "balanced"
                                  "env" {}
                                  "expired-instance-restart-rate" 0.1
                                  "grace-period-secs" 30
                                  "health-check-url" "/status"
                                  "idle-timeout-mins" 30
                                  "instance-expiry-mins" 7200 ; 5 days
                                  "jitter-threshold" 0.5
                                  "max-instances" 500
                                  "max-queue-length" 1000000
                                  "metadata" {}
                                  "min-instances" 1
                                  "permitted-user" "*"
                                  "restart-backoff-factor" 2
                                  "scale-down-factor" 0.001
                                  "scale-factor" 1
                                  "scale-up-factor" 0.1}
   :statsd :disabled
   :thread-stack-state-refresh-interval-ms 600000 ; 10 minutes
   :work-stealing {:offer-help-interval-ms 100
                   :reserve-timeout-ms 1000}
   :zookeeper {:base-path "/waiter"
               :curator-retry-policy {:base-sleep-time-ms 100
                                      :max-retries 10
                                      :max-sleep-time-ms 120000}
               :discovery-relative-path "discovery"
               :gc-relative-path "gc-state"
               :leader-latch-relative-path "leader-latch"
               :mutex-timeout-ms 1000}})

(defn deep-merge-settings
  "Merges the two settings maps, with special handling for maps with :kind"
  [map-1 map-2]
  (merge-with
    (fn [x y]
      (if (and (map? x) (map? y))
        (let [x-kind (:kind x)
              y-kind (:kind y)]
          (if (and x-kind y-kind)
            (let [kind y-kind
                  x-sub-map (get x kind)
                  y-sub-map (get y kind)]
              (-> x
                  (merge y)
                  (assoc kind (merge x-sub-map y-sub-map))))
            (deep-merge-settings x y)))
        y))
    map-1 map-2))

(defn load-settings
  [config-file git-version]
  (deep-merge-settings
    settings-defaults
    (assoc
      (load-settings-file config-file)
      :git-version git-version)))

